File: tests/test_files/go/service.go
package main

type service struct{ db readonlyDB }

func runService(write bool) {
	svc := newService(write)
	svc.foo()
	svc.bar()
	svc.save()
}

func newService(write bool) *service {
	if write {
		return &service{db: ReadWriteDB{}}
	}
	return &service{db: ReadOnlyDB{}}
}

func (svc *service) foo() {
	svc.db.query("foo")
}

func (svc *service) bar() {
	svc.db.query("bar")
}

func (svc *service) save() {
	if w, ok := svc.db.(writeableDB); ok {
		w.update("save")
	}
}

File: tests/test_files/rust/service.rs
use crate::db::*;

pub struct Service {
    ro_db: ReadOnlyDb,
    rw_db: ReadWriteDb,
}

pub fn run() {
    let mut svc = Service::new();
    svc.foo();
    svc.bar();
    svc.save();
}

impl Service {
    pub fn new() -> Service {
        Service {
            ro_db: ReadOnlyDb {},
            rw_db: ReadWriteDb {},
        }
    }

    pub fn foo(&self) {
        self.ro_db.query("foo");
    }

    pub fn bar(&mut self) {
        self.ro_db.query("bar");
        self.rw_db.query("bar");
    }

    pub fn save(&mut self) {
        self.rw_db.update("save");
    }
}
